pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -313,6 +334,43 @@ private synchronized void completeConnect(AuthData clientData) throws PulsarClie
         }
     }
 
+    private void handleBrokerConnected(DirectProxyHandler directProxyHandler, CommandConnected connected) {
+        checkState(ctx.executor().inEventLoop(), "This method should be called in the event loop");
+        if (state == State.ProxyConnectingToBroker && ctx.channel().isOpen() && this.directProxyHandler == null) {
+            this.directProxyHandler = directProxyHandler;
+            state = State.ProxyConnectionToBroker;
+            int maxMessageSize =
+                    connected.hasMaxMessageSize() ? connected.getMaxMessageSize() : Commands.INVALID_MAX_MESSAGE_SIZE;
+            ctx.writeAndFlush(Commands.newConnected(connected.getProtocolVersion(), maxMessageSize))
+                    .addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+        } else {
+            LOG.warn("[{}] Channel is {}. ProxyConnection is in {}. "
+                            + "Closing connection to broker '{}'.",
+                    remoteAddress, ctx.channel().isOpen() ? "open" : "already closed",
+                    state != State.ProxyConnectingToBroker ? "invalid state " + state : "state " + state,
+                    proxyToBrokerUrl);
+            directProxyHandler.close();
```

Comment:
What is the purpose of closing this here? I'd expect the `channelInactive` method to close it. Is it an optimization to close the channel as early as possible? If so, should we set it to `null`?

Reply:
`directProxyHandler` is the method parameter in this case. it's not set to any field at this point. The instance will only be set to the field if the state is what is expected (`state == State.ProxyConnectingToBroker && ctx.channel().isOpen() && this.directProxyHandler == null`).

Reply:
the else branch would get executed in the case if the incoming connection gets closed while the broker connection is being created. That's the race that this PR intends to fix.

Reply:
Thanks, I missed that nuance.